server/src/main/java/org/opensearch/action/search/QueryPhaseResultConsumer.java

```diff
@@ -382,129 +387,62 @@ long ramBytesUsedQueryResult(QuerySearchResult result) {
-            if (circuitBreakerBytes > 0) {
-                circuitBreaker.addWithoutBreaking(-circuitBreakerBytes);
-                circuitBreakerBytes = 0;
+            if (result.isNull()) {
+                SearchShardTarget target = result.getSearchShardTarget();
+                emptyResults.add(new SearchShard(target.getClusterAlias(), target.getShardId()));
+                return true;
             }
-        }
-
-        private void onAfterMerge(MergeTask task, MergeResult newResult, long estimatedSize) {
-            synchronized (this) {
-                runningTask.compareAndSet(task, null);
-                if (hasFailure()) {
-                    task.cancel();
-                    return;
+            // Check circuit breaker before consuming
+            if (hasAggs) {
+                long aggsSize = ramBytesUsedQueryResult(result);
+                try {
+                    addEstimateAndMaybeBreak(aggsSize);
+                    aggsCurrentBufferSize += aggsSize;
+                } catch (CircuitBreakingException e) {
+                    onMergeFailure(e);
+                    return true;
                 }
-                mergeResult = newResult;
-                if (hasAggs) {
-                    // Update the circuit breaker to remove the size of the source aggregations
-                    // and replace the estimation with the serialized size of the newly reduced result.
-                    long newSize = mergeResult.estimatedSize - estimatedSize;
-                    addWithoutBreaking(newSize);
-                    logger.trace(
-                        "aggs partial reduction [{}->{}] max [{}]",
-                        estimatedSize,
-                        mergeResult.estimatedSize,
-                        maxAggsCurrentBufferSize
-                    );
-                }
-                task.consumeListener();
             }
+            // Process non-empty results
+            int size = buffer.size() + (hasPartialReduce ? 1 : 0);
+            if (size >= batchReduceSize) {
+                hasPartialReduce = true;
+                // the callback must wait for the new merge task to complete to maintain proper result processing order
+                QuerySearchResult[] clone = buffer.toArray(QuerySearchResult[]::new);
+                MergeTask task = new MergeTask(clone, aggsCurrentBufferSize, new ArrayList<>(emptyResults), callback);
+                aggsCurrentBufferSize = 0;
+                buffer.clear();
+                emptyResults.clear();
+                queue.add(task);
+                tryExecuteNext();
+                buffer.add(result);
+                return false; // callback will be run by merge task
+            }
+            buffer.add(result);
+            return true;
         }
 
         private void tryExecuteNext() {
             final MergeTask task;
             synchronized (this) {
                 if (hasFailure()) {
-                    clearPendingMerges(null);
```

Comment:
We had kept this because keeping it will make sure we are not hogging any memory by pooling partialReduce tasks in case of failure.

Reply:
Now the flow is 
Failure -> Handle failure and clear all resources -> Set failure reference -> `hasFailure` returns true
So we can just return early.